firedrake/slate/slac/compiler.py

```diff
@@ -0,0 +1,324 @@
+
+    if builder.needs_cell_facets:
+        args.append(ast.Decl("char *", cellfacetsym))
+
+    macro_kernel_name = "compile_slate"
+    kernel_ast, oriented = builder.construct_ast(name=macro_kernel_name,
+                                                 args=args,
+                                                 statements=ast.Block(statements))
+
+    inc.append(petsc.get_petsc_dir() + '/include/eigen3/')
+    op2kernel = op2.Kernel(kernel_ast, macro_kernel_name, cpp=True, include_dirs=inc,
+                           headers=['#include <Eigen/Dense>', '#define restrict __restrict'])
+
+    assert len(slate_expr.ufl_domains()) == 1
+    kinfo = KernelInfo(kernel=op2kernel,
+                       integral_type="cell",
+                       oriented=oriented,
+                       subdomain_id="otherwise",
+                       domain_number=0,
+                       coefficient_map=range(len(slate_expr.coefficients())),
+                       needs_cell_facets=builder.needs_cell_facets)
+    idx = tuple([0]*slate_expr.rank)
+
+    return (SplitKernel(idx, kinfo),)
+
+
+def auxiliary_information(builder):
+    """This function generates any auxiliary information regarding special handling of
+    expressions that do not have any integral forms or subkernels associated with it.
+
+    :arg builder: a :class:`SlateKernelBuilder` object that contains all the necessary
+                  temporary and expression information.
+
+    Returns: a mapping of the form ``{aux_node: aux_temp}``, where `aux_node` is an
+             already assembled data-object provided as a `ufl.Coefficient` and `aux_temp`
+             is the corresponding temporary.
+
+             a list of auxiliary statements are returned that contain temporary declarations
+             and any code-blocks needed to evaluate the expression.
+    """
+    aux_temps = {}
+    aux_statements = []
+    for i, exp in enumerate(builder.aux_exprs):
+        if isinstance(exp, Action):
+            acting_tensor = exp.tensor
+            acting_coefficient = exp._acting_coefficient
+            assert isinstance(acting_coefficient, Coefficient)
+
+            temp = ast.Symbol("C%d" % i)
+            # TODO: think about the shape of this temporary
+            cshape = acting_tensor.shape[1]
+            aux_statements.append(ast.Decl(eigen_matrixbase_type(shape=(cshape,)), temp))
+            aux_statements.append(ast.FlatBlock("%s.setZero();\n" % temp))
+            itersym = ast.Symbol("i1")
+
+            action_loop = ast.For(ast.Decl("unsigned int", itersym, init=0),
+                                  ast.Less(itersym, cshape),
+                                  ast.Incr(itersym, 1),
+                                  ast.Assign(ast.Symbol(temp, rank=(itersym,)),
+                                             ast.Symbol(builder.coefficient_map()[acting_coefficient],
+                                                        rank=(itersym, 0))))
```

Comment:
Since the removal of `flatten=True`, this is now the wrong unpacking in the case of VectorFunctionSpace/TensorFunctionSpace.  Since the data are now packed as `coefficient[node_extent][dof_extent]`, rather than `coefficient[node_extent*dof_extent][1]`.